### ISSUE TITLE
Cache failed builds

### ISSUE THREAD

**Issue by cs-shadow** [cs-shadow/buildstream#80]
In GitLab by @sstriker on Aug 25, 2017, 10:01

When inputs don't change, there is no reason to expect a build to succeed a second time.  Currently when executing `bst build` on a bst that previously failed will retry the build.

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 25, 2017, 18:07

> Dear user, your project produces the artifact `gtk+.bd5k8sd3fs49fse55.....` which is known to fail. If you want to try building it anyway, try again with `bst build --force ...`


Interesting indeed.


We could generate artifacts for failed builds and share those failures with peers too. A user trying to build something which failed could just end up downloading a log file and seeing what went wrong.

And of course I guess you have to be allowed to bypass it and build it anyway in case you are interested in debugging it in a shell ?

**Comment by cs-shadow** [comment]
In GitLab by @sstriker on Aug 25, 2017, 20:07

Thanks for expanding.

Yes indeed, it should be overridable as you sketch.

**Comment by cs-shadow** [comment]
In GitLab by @hashpling on Aug 29, 2017, 09:12

Even running in a sandbox, there are events where you might expect the a build to succeed when run a second time. One example, which has happened more than once, is running a build in Docker with the maximum allocated memory set too low. Re-running the identical build with sufficient memory can be expected to succeed.

**Comment by cs-shadow** [comment]
In GitLab by @sstriker on Aug 29, 2017, 17:06

I would still be ok with having the explicitly retry those builds than to re-run them by default.  Whether we have a global override switch which always tries to build is another question.

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Mar 30, 2018, 04:50

Note that @hashpling's concern *might* be addressable with some additional load balancing or configuration, this was raised separately in #185 and should probably be discussed there instead, maybe there are some creative solutions we can come up with.

An additional note I have to add to the subject of caching failed builds, is related to how we handle local build failures.

I think that if we add functionality to:
* Cache build directories in artifacts for later use (a feature which has been developed and should land as a part of #311)
* Cache failed build artifacts, which should include the build directory in it's failed state along with anything which *might* have been produced in `%{install-root}`
* Allow sharing of failed build artifacts
* Allow debugging of a failed build state, as a feature of `bst shell`, and as a possibility for force retrying in `bst build`

Then we should have a single story for retrying a failed build and not separate ones depending on whether it failed locally or remotely.

Essentially I think that when we cache a failed build artifact, we should immediately delete the local build directory, as it will be in the cached artifact anyway. This will solve the issue of unmanaged, ever growing build directories in `~/.cache/buildstream/build` at the same time.

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Apr 23, 2018, 16:06

The caches currently lack any interface to get metadata
independent of the contents of the artifact.

Currently the first we know about an artifact's contents
is when we are staging it to assemble a dependent.

If we were to just add failed builds to the cache
and build result metadata to the artifact
then we would have to fail the first time the artifact is extracted.

This would likely be an improvement in responsiveness
compared with attempting a build that will fail,
since extracting would be less work than building,
but if we knew sooner we could opt to fail before starting any builds
or would know to enqueue a rebuild.


If we were to make the queue more dynamic
we could solve the early-failing or rebuilding issue.

We could enqueue extract jobs before build jobs,
staging from the pre-extracted artifacts,
and if it's a failed build artifact then enqueue a rebuild
before dependent builds of the failed build artifact.

Alternatively we could during the staging stage of a build
if the artifact being extracted was a failed build
queue a rebuild job of both the failed build and the current build
and terminate the current build job.


Instead of making the queue more dynamic
we could make the cache able to distinguish successful from failed builds,
and keep roughly the same architecture.

If we care about cache compatibility between versions of buildstream
we could bodge the fact that it's a failed build into the cache key
(either by extending the name with a token or adding it to the computation)
and when looking in the cache for an artifact, check both keys.

Unless it's an obvious and parseable part of the name
this wouldn't be accessible to cache cleanup policy logic.

An alternative would be to maintain a separate cache for failures,
though this would require changing a lot of code paths to expect more caches
or to have a wrapper cache that operates on both,
though this could require changes to the
this wouldn't be accessible to cache cleanup policy logic.

An alternative would be to maintain a separate cache for failures,
though this would require changing a lot of code paths to expect more caches
or to have a wrapper cache that operates on both,
though this could require changes to the
this wouldn't be accessible to cache cleanup policy logic.

An alternative would be to maintain a separate cache for failures,
though this would require changing a lot of code paths to expect more caches
or to have a wrapper cache that operates on both,
though this could require changes to the API.

A more long-term alternative would be to include metadata alongside content,
but while we could probably do that fairly easy in the tar cache
by adding a nice machine parseable file alongside,
I've not yet any suggestions for how best to handle it with ostree yet.

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Apr 23, 2018, 16:48

I'm going to have a deeper look at options for metadata in the caches and put together a proposal for what to do, then barring objections, begin implementing it.
I just thought my write-up would be useful if anyone has time to comment.

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Apr 24, 2018, 16:51

OSTree has a couple of ways of handling metadata.

There's "detached metadata" for metadata describing a commit
that may change over time (which could be useful for tracking usage),
or the commit may include as much additional metadata as wanted,
though affects the OSTree cache ID this shouldn't be an issue
since the Buildstream Element cache ID wouldn't change.

By extending the Artifact Cache API to include a method for getting the metadata
the Pipeline planner could, in the same way that it's told to ignore the cache,
be told to ignore cached failures and rebuild, or decide it's not buildable.

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Apr 26, 2018, 14:40

The Google Content-Addressable Store itself operates on a lower level than artifact caches.
It doesn't have a native mechanism for metadata, presumably because you could instead build your own mechanism with a top-level directory containing a subdirectory of the contents and a separate one for the metadata.

Given the goal is to be able to fetch metadata to make a decision on whether a build was a failure it's unfortunate that we can't use the `GetTree` method of the Remote Execution API for this, since from my understanding it will download all the tree data in an arbitrary order rather than just a requested subtree.

In terms of the inputs and outputs, the Artifact Cache seems closer in concept to the Remote Action Cache, the difference being that our caches are provided with the key rather than computing it from the build description.
A Remote Action Cache `ActionResult` includes the `exit_code` metadata which could encode whether a build was a success or a failure.

As a result I don't think we can so comfortably include arbitrary metadata in the artifact cache, but we could include a result, which could either be an enum from wrapping commands we send to a Remote Action Cache service, or an arbitrary exit code.

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Apr 27, 2018, 06:20

Hmmm, ok so at first sight of this, I find the description is misleading.

The goal of caching failed builds is indeed partly to avoid building things which we can safely predict will fail, but I feel that the bigger motivation here, is to allow sharing of the failed build artifacts through a cache server.

I.e.: When a build fails on a CI server, currently all you have is the logs; instead, we should be able to download the related artifacts such that we can run a `bst shell` on them and debug what went on remotely, locally.

@richardmaw\-codethink: I think that you are overthinking things by desiring to get the `ArtifactCache` to perform partial extractions.

Rather, I think that we might benefit from coupling the artifact extraction with the artifact creation:
* If we built an artifact, there is a *very* high likelyhood that we will later need to extract it
  * Currently when a build completes and separate builds blocking on that build begin simultaneously, we have separate extractions which occur in parallel, and the winner replaces the extract dir while the loser's work in extraction is discarded. This is an overhead which we would reduce by forcing the extraction to occur while caching the artifact
* If we downloaded an artifact, the same cost is likely also incurred when racing to extract the artifacts at build time
* If it is a failed artifact which we build
  * The likelyhood that we want to shell into it is high, this will anyway require an extraction
* If it is a failed artifact which we downloaded
  * Here is one place where we *might* want a bit more optimization, the likelyhood that we want to extract it is a bit lower

However, I should further note that:
* With `OSTree`, checkouts of individual full artifacts should be very quick, and not require additional storage, they will be done with hardlinks.
* With the upcoming `CAS`, I expect checkouts to be similarly optimized

In this light, I think that things will be faster overall if we extract at cache time, and that the overhead for the *very few artifacts* which we did not need to really extract is sufficiently negligent to not merit overengineering the artifact caches for partial extractions.

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Apr 27, 2018, 07:25

@richardmaw\-codethink I think I have an idea about this.

From what I understand, a tricky part of implementing this is that we assume:
* We are running `bst build target.bst`
* We have downloaded a failed `dependency.bst` artifact

In this case, we probably want to handle things at the scheduler `Queue` level.

* Elements pass through queues, such that they go through `PullQueue` before ending up in `BuildQueue`
* When an element arrives in the `BuildQueue`, but it is already *cached*, we currently *skip* the element
* Instead of *skipping* the element, the `BuildQueue` should mark the element as *failed*, and go through the regular process of handling a failed build.
* This also requires that we have a new `Element._failed()` state which can be observed from the Queues, in addition to other exposed states like `Element._cached()`, this will be fairly symmetric with other existing code, and should be resolved withing `Element._update_state()`.

As I mentioned in https://gitlab.com/BuildStream/buildstream/issues/76#note_65830031, with this we should have a uniform way of debugging failed builds, and this should probably be the default mode of operation.

I would expect that if I have downloaded a failed build, I will be presented with an interactive prompt, and handle things in exactly the same way as if I had failed to build the artifact myself.

Note that reconstructing a build environment from a failed build artifact also depends on the caching of the build tree, which is being worked on in #311 and is in turn, also blocking on #376.

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Apr 27, 2018, 17:00

A lot of my ramblings are irrelevant given an artifact's contents is extracted and retained.

I'd assumed that extract happened directly into the staging area, not noticing that splitting is handled after being extracted from the cache rather than before they were put in.

This simplifies things for me, though it feels weird that we're effectively keeping two caches.


From my reading of the job planner logic it could also omit including the element in the queue in the first place, though I'd have to double check.

While a cached failure would let us skip the build of the failed element itself, we could also skip all the other builds before we start them if it was our policy to treat a cached failure as a build fail.

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Apr 28, 2018, 12:06

> This simplifies things for me, though it feels weird that we're effectively keeping two caches.

Yeah it seems weird I know, it's not as bad as it is with tarball based artifact caches though; since all the unpacked files in the `extract/` directory are essentially hardlinks to deduplicated objects in the `OSTree` repo (so it doesnt cost immense amounts of extra disk space, asides from directory nodes).

> While a cached failure would let us skip the build of the failed element itself, we could also skip all the other builds before we start them if it was our policy to treat a cached failure as a build fail.

I think we discussed in the past that it might be interesting to have a force retry option for retrying a cached failed build so we might try to just have the Queue skip it and mark it failed by default and have retry happen with an explicit option; the result would be similar, we would end up running the scheduler but it might just be a noop.

Note that because of the [remote execution plan](https://mail.gnome.org/archives/buildstream-list/2018-April/msg00006.html), the CAS artifact cache will no longer support the summaries like OSTree does, so if the cached failed build is going to be downloaded, there is not really going to be any way to predict this at startup time. In fact even without that, we dont know if an artifact which we will download is failed or not until we unpack it.

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on May 22, 2018, 14:25

assigned to @richardmaw-codethink

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on May 22, 2018, 16:45

My WIP is currently able to store failed builds in the cache and treats a cached fail like it had just tried and failed to build.
Potentially it should include being able to be instructed to retry a failed build once, policy for what to do with a cached fail, and speculatively extracting artifacts before it's accepted, but it's probably more useful to rebase and get the WIP merge request up at this point.

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on May 23, 2018, 15:33

Note my comment here: https://gitlab.com/BuildStream/buildstream/issues/402#note_74602072

This issue should ideally land after fixing #21, and we should stop leaving failed build directories on disk, otherwise, we will need a new issue to implement this cleanly once #21 is ready to land (although I admit there seems to be not much utility to this issue if you dont have the source tree cached and ready to reconstruct a build environment to debug).

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on May 23, 2018, 15:44

Something new comes to mind: when reconstructing a failed build, we may need to be *strict* about the dependency artifacts we stage, even in *non strict mode*. More than this even: we will want to ensure that we use exactly the same artifacts which were used in the failing build, even if the build failed in *non strict mode*.

* this may require additional artifact metadata in the failed artifact
* the possibility of failing to reconstruct is minimal, because non strict build artifacts are never shared with peers on an artifact server

**Comment by cs-shadow** [comment]
In GitLab by @toscalix on May 23, 2018, 15:47

marked this issue as related to #402

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on May 25, 2018, 14:51

mentioned in merge request !475

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Jun 1, 2018, 15:49

MR has been updated to the point it may have an initial review.

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Jun 5, 2018, 14:12

mentioned in commit b66e8dd09023db91d852b4c46389d68e264629aa

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Jun 7, 2018, 13:03

mentioned in commit 36ddbae22c00e27ab592e71ecc302cca5d40985a

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Jun 7, 2018, 14:24

!475 currently enables the caching of failed builds, but does not implement pushing of failed builds to a common cache server, which was one of the key objectives of being able to cache failed builds.

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Jun 13, 2018, 15:36

mentioned in commit b7ee1c9629663a953d3d8a0761f68c5d1680d0f3

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Jun 15, 2018, 09:47

mentioned in commit 43dfbf3f49a606f4ee280da38cfa0a8e0fe66108

**Comment by cs-shadow** [comment]
In GitLab by @LaurenceUrhegyi on Jun 19, 2018, 11:56

Moved into 'Review' column and prodded @tristanvb for a review of https://gitlab.com/BuildStream/buildstream/merge_requests/475

**Comment by cs-shadow** [comment]
In GitLab by @toscalix on Jun 24, 2018, 14:35

Status Review does not mean the the code in the ticket is being review by the maintainers. It means that somebody is checking that the acceptance criteria is being met. The status review works for issues involving code and those that does not. Back to ~Doing

**Comment by cs-shadow** [comment]
In GitLab by @LaurenceUrhegyi on Jun 24, 2018, 15:19

Ah, ok, thanks for the clarification. Now I know :)

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Jun 26, 2018, 10:29

https://gitlab.com/BuildStream/buildstream/merge_requests/475 has had feedback.

I've had some more urgent work jump to the beginning of my queue so I haven't had time to look at it.
It's going to be a few days before I can.

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Jul 3, 2018, 09:40

I expect to be able to look at this again early next week.

**Comment by cs-shadow** [comment]
In GitLab by @LaurenceUrhegyi on Jul 4, 2018, 17:14

Moved to 'ToDo' and un-assigned for now.

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Jul 23, 2018, 11:01

mentioned in merge request !504

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Jul 26, 2018, 11:21

I see that there has been some movement on !475 for this, and something came to mind.

Now that @knownexus has landed and closed #21, this unblocks a lot of work related to using the cached build trees, particularly the parts of this issue which make cached failed builds useful: the ability to download a failed build and debug it locally.

In light of this unblocking, we have a choice to make:
* We can continue with !475 as is, without making use of the buildtree, and accumulate some deficit to be paid back during the 1.3.x cycle:
  * We'll have to work on reconstructing the build environment from an artifact including it's build tree as a separate activity, and file a separate issue for that
  * We'll have to ensure that we no longer leak build directories in ~/.cache/buildstream/build, and have a single codepath for debugging a failed build, regardless of whether the failure occurred locally and we're debugging in an interactive shell or whether the artifact failed remotely. This would be either a part of the afore mentioned activity, or yet again in another ticket.
* Or, we can implement the whole thing as a part of this activity
  * It seems sensible to me, since as far as I can see there is no work which blocks on having failed cached builds without actually using the buildtree and being able to debug them

I can go either way, and would like feedback from others on how they would like to proceed here.

**Comment by cs-shadow** [comment]
In GitLab by @toscalix on Jul 26, 2018, 12:41

unassigned @richardmaw-codethink

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Jul 27, 2018, 16:44

mentioned in commit 236f9958b1f85e9eb78ccbce43ac70cd8b5cebaa

**Comment by cs-shadow** [comment]
In GitLab by @richardmaw-codethink on Jul 31, 2018, 16:34

mentioned in commit e8cd43dc50c41c44bf29d68e4ce947182de5d848

**Comment by cs-shadow** [comment]
In GitLab by @knownexus on Jul 31, 2018, 17:17

closed via merge request !475

**Comment by cs-shadow** [comment]
In GitLab by @knownexus on Jul 31, 2018, 17:17

mentioned in commit 4fc1f5d1e82ac5e2f22440e648a901a83f8eaa61

**Comment by cs-shadow** [comment]
In GitLab by @LaurenceUrhegyi on Aug 1, 2018, 11:52

marked this issue as related to #536

**Comment by cs-shadow** [comment]
In GitLab by @LaurenceUrhegyi on Aug 1, 2018, 11:56

marked this issue as related to #534

**Comment by cs-shadow** [comment]
In GitLab by @LaurenceUrhegyi on Aug 1, 2018, 12:01

Sadly you don't seem to have been given any feedback here before this was merged and then this issue closed off.

So AFAICT we now have two tickets raised which are the 'technical debt' of merging this, these are: https://gitlab.com/BuildStream/buildstream/issues/534 and https://gitlab.com/BuildStream/buildstream/issues/536.

I'll see if we need any more tickets raised to cover off what you outline in your comments above.

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 2, 2018, 08:38

Update, I have filed #539 which more accurately describes the technical debt resulting from this feature. #539 is a subset of #536, if @LaurenceUrhegyi wants to keep track of multiple things in #536 and keeps that issue open is up to him.

**Comment by cs-shadow** [comment]
In GitLab by @knownexus on Aug 2, 2018, 14:15

removed the relation with #534
